File: src/main/java/edu/cornell/cs/apl/viaduct/util/JoinSemiLattice.java
package edu.cornell.cs.apl.viaduct.util;

/** A set that supports binary least upper bounds. */
public interface JoinSemiLattice<T extends JoinSemiLattice<T>> extends PartialOrder<T> {
  /** Least upper bound of {@code this} and {@code that}. */
  T join(T that);
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/JoinSemiLattice.kt
package io.github.apl_cornell.viaduct.algebra

/** A set that supports binary least upper bounds. */
interface JoinSemiLattice<T : JoinSemiLattice<T>> {
    /** The least upper bound of `this` and [that]. */
    infix fun join(that: T): T
}